### DISCUSSION TITLE
Provide global HttpClient to use_future

### DISCUSSION THREAD

**Discussion by patrick-dedication** [DioxusLabs/dioxus#832] [Q&A] [answered]
Hi,

first off, thanks for making Rust web so great.
I'm trying to provide a global HttpClient so I can use it in my http hooks.
I fear my implementation is aweful, because I clone the HttpClient on every render.
How can I improve?
```
#[allow(non_snake_case)]
pub fn App(cx: Scope) -> Element {
    use_context_provider(cx, || FetchHttpClient::new("http://127.0.0.1:8000"));

    let slides = use_slides(cx);

    cx.render(match slides.value() {
        Some(slides) => rsx! { Slideshow { slides: slides } },
        None => rsx! { div { "Loading slides..." } },
    })
}
```

```
pub fn use_slides(cx: &ScopeState) -> &UseFuture<Slides> {
    let http_client: FetchHttpClient = use_context::<FetchHttpClient>(cx).unwrap().clone();
    use_future(cx, (), |_| async move { get_slides(&http_client).await })
}
```

Thanks in advance!

**Comment by ealmloff** [comment, marked as answer]
You can use a [Lazy static](https://docs.rs/once_cell/latest/once_cell/sync/struct.Lazy.html) variable to initialize FetchHttpClient:
```rust
static CLIENT: once_cell::sync::Lazy = once_cell::sync::Lazy(|| FetchHttpClient::new("http://127.0.0.1:8000"));

#[allow(non_snake_case)]
pub fn App(cx: Scope) -> Element {
    let slides = use_slides(cx);

    cx.render(match slides.value() {
        Some(slides) => rsx! { Slideshow { slides: slides } },
        None => rsx! { div { "Loading slides..." } },
    })
}
```
```rust
pub fn use_slides(cx: &ScopeState) -> &UseFuture<Slides> {
    use_future(cx, (), |_| async move { get_slides(&CLIENT).await })
}
```

**Reply by patrick-dedication** [reply]
Thanks for the fast response.
That is probably the right thing to do in my case.
Still I wonder, if there is a pattern in Dioxus to provide some semi global struct that I can change depending on the context, i.e. for Tests.
We do that in our React projects, there we would render the component wrapped in a context provider.

**Reply by ealmloff** [reply]
The original code you provided is Dioxus' solution to semi-global state, but you can improve the code in several ways: You can wrap the state in a Rc or Arc to make the state cheaply clonable and either use the dependencies to avoid explicitly cloning the state every time or move the consume context into the future code.
```rust
#[allow(non_snake_case)]
pub fn App(cx: Scope) -> Element {
    use_context_provider(cx, || Rc::new(FetchHttpClient::new("http://127.0.0.1:8000")));
    let slides = use_slides(cx);

    cx.render(match slides.value() {
        Some(slides) => rsx! { Slideshow { slides: slides } },
        None => rsx! { div { "Loading slides..." } },
    })
}
```
```rust
pub fn use_slides(cx: &ScopeState) -> &UseFuture<Slides> {
    use_future(cx, (), |_| {
        let http_client: Rc<FetchHttpClient> = cx.consume_context().unwrap();
        async move { get_slides(&).await }
    })
}
```

**Reply by patrick-dedication** [reply]
Thanks a lot!

**Reply by patrick-dedication** [reply]
Hi, I finally got around to try the suggestion.
It turns out, I can't use `cx.consume_context()` inside the FnOnce passed to `use_future`.
The lifetime of `cx` must be `'static` to do so, is there a way to solve this?
The full code is here in case you need more context: https://github.com/patrick-dedication/rust-web-slideshow

**Reply by ealmloff** [reply]
The closure is not required to be 'static, but the future is:

Here is a working version like the above example
```rust
use std::rc::Rc;

pub fn use_slides(cx: &ScopeState) -> &UseFuture<()> {
    use_future(cx, (), |_| {
        // cloned before it is moved into the 'static async block
        let http_client: Rc<i32> = cx.consume_context().unwrap();
        async move { println!("{}", &http_client) }
    })
}
```

Here is a similar version that will output the error you received
```rust
pub fn use_slides_broken(cx: &ScopeState) -> &UseFuture<()> {
    use_future(cx, (), |_| async move {
        // because this is inside the 'static async block, cx is 'static
        let http_client: Rc<i32> = cx.consume_context().unwrap();
        println!("{}", &http_client)
    })
}
```

**Reply by patrick-dedication** [reply]
Thanks again, and thanks for taking the time to explain :)
